rfcs/0016-error-handling.md

```diff
@@ -0,0 +1,379 @@
+- `should_panic` (without values) will assert if the test reverts with a compiler generated revert code. The difference between the existing `should_revert` and `should_panic` is, that the former can assert _any_ revert and can specify the expected revert id, while the latter can only assert that the revert code was compiler generated (>= 0xffff_ffff_0000_0000).
+- `panic_msg = "<regex>"` asserts that the error message related to the `panic`king matches the provided `regex`,
+- `panic_value = "<regex>"` asserts that the logged error value related to the `panic`king matches the provided `regex`. The values will be displayed as decoded logs are currently displayed. The `regex` must match the displayed representation. E.g., if the logged error value is `MyEnumVariant(MyStruct { x: 42 })` we can assert it with any of the following regexes, depending on the required strictness: `42`, `x..42`, `MyStruct.*x.*42`.
+
+Similarly, the SDKs will recognize reverts coming from `panic`king and display their ABI error information accordingly.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Adapt compilation of `panic` to avoid structural changes in the `std`
+
+Currently, the implementation of `panic` wraps the argument into an `encode` call to be able to log it. This makes it impossible to use `panic` even with constant string slice arguments in `std` modules like `ops`, because of circular dependencies with the `codec` module. Rather then restructuring the standard library, we will change the compiler to recognize constant string slices as a special case and do not `encode` them, which will remove the need for `encode`ing and thus the circular dependency.
+
+## Extend the `#[test]` attribute
+
+Adding the proposed new arguments to the `#[test]` attribute is a straightforward reuse of the existing compiler infrastructure for declaring attributes, and does not require any changes in the compiler.
+
+## Adapt existing reusable libraries
+
+The majority of the work will go into adapting the `std` and other libraries to follow the guidelines presented in [Error handling in reusable libraries](#error-handling-in-reusable-libraries).
+
+All the existing enums used for error reporting, like, e.g. `standards::src5::AccessError` will be annotated as `#[error_type]`s.
+
+As an example of a typical adaptation of a _reusable library_, let's consider the `sway_libs::ownership` module. Calling `only_owner` as an entry guard in every `ownership`'s function as it is now, would result in an ABI error always located in the `std::revert::require` function. This is against the guidelines, as it provides an error location unnecessary separated through two levels of indirection from the actual guard invocation point.
+
+Instead, we should introduce a private `is_owner` function and panic in every `sway_libs::ownership`'s function separately:
+```
+if !is_owner() {
+    panic AccessError::NotOwner;
+}
+```
```

Comment:
There's an issue we anticipated that `panic` would not be adequately composable because you can't bubble up the source of the panic to where it is actually relevant.

Maybe instead of copy and paste we should solve that so that writing a `require` function that uses `panic` is actually possible?